### DISCUSSION TITLE
Variables, constants and translations

### DISCUSSION THREAD

**Discussion by NKL-aatu** [PrivateBin/PrivateBin#1022] [General]
Hi.

Developing a template. Just a few quick questions:

1. In the default template there are variables like $isCpct, $isDark and $isPage and constants like $SYNTAXHIGHLIGHTING and $MARKDOWN. Have I understood correctly that those variables are used to refer browser/UA settings and constants are used to refer current PrivateBin installation settings in the configuration file?
2. What are $isCpct, $isDark and $isPage variables? My guess is $isCpct refers to mobile view and $isDark should read UA's darkmode setting (but doesn't work at least on my phone)... I haven't been able to find an instance where $isPage condition gets triggered.
3. Are current paste's settings and content all fetched with JavaScript? So no PHP variable or constant in the template file holds this data?
4. Does PrivateBin currently support adding organization specific translations to UI texts? At the moment I'm using a less pretty workaround: an extra find and replace function that's always called before actual translation functions.

**Comment by elrido** [comment]
> 1. In the default template there are variables like $isCpct, $isDark and $isPage and constants like $SYNTAXHIGHLIGHTING and $MARKDOWN. Have I understood correctly that those variables are used to refer browser/UA settings and constants are used to refer current PrivateBin installation settings in the configuration file?
> 2. What are $isCpct, $isDark and $isPage variables? My guess is $isCpct refers to mobile view and $isDark should read UA's darkmode setting (but doesn't work at least on my phone)... I haven't been able to find an instance where $isPage condition gets triggered.

The camelCased ones are specific to the bootstrap templates and is how the single file re-arranges it's content to render the 5 different variants of the bootstrap and bootstrap-* templates, as [depicted on our screenshots page](https://privatebin.info/screenshots.html). They are calculated at the top of that template file and are based on the template configuration setting, not dynamic based on user agent:
https://github.com/PrivateBin/PrivateBin/blob/910675c74b3a4c63eb0dc963a714d4041cf0eb66/tpl/bootstrap.php#L3-L5

The ALLCAPS ones are assigned to the View and present in all templates. They represent configuration values, or are based on dynamic properties of the server or client, etc.

https://github.com/PrivateBin/PrivateBin/blob/910675c74b3a4c63eb0dc963a714d4041cf0eb66/lib/Controller.php#L383-L417

> 3. Are current paste's settings and content all fetched with JavaScript? So no PHP variable or constant in the template file holds this data?

The paste's and comment's settings are all fetched from the API. There are some server settings that get injected into data-attributes, if present, i.e.:
https://github.com/PrivateBin/PrivateBin/blob/910675c74b3a4c63eb0dc963a714d4041cf0eb66/tpl/bootstrap.php#L525

The JS client enables or disables bits parts of the static HTML as needed, if present. For a custom template, the HTML structure is pretty flexible, just the key id-attributes have to exist and relevant HTML elements within them, e.g. a textarea with id "message":

https://github.com/PrivateBin/PrivateBin/blob/910675c74b3a4c63eb0dc963a714d4041cf0eb66/tpl/bootstrap.php#L564

> 4. Does PrivateBin currently support adding organization specific translations to UI texts? At the moment I'm using a less pretty workaround: an extra find and replace function that's always called before actual translation functions.

Right now only the [name option](https://github.com/PrivateBin/PrivateBin/wiki/Configuration#name) can be used to replace the "PrivateBin" labels in the translations. For any other adjustments you would have to change the relevant messages in all the translation files. During upgrades you would then have to add any newly translated strings on the bottom of your modified translation files - we append new messages there to make it easy to spot new and untranslated messages for the translaters (less relevant now that most of them use the Crowdin UI, where the untranslated messages are automatically flagged).

And of course you can add other, untranslated messages in any place of your new template - many instance operators change the footer text blurb, for example.

**Reply by NKL-aatu** [reply]
Thank you so much! I would have never guessed :D

So now I know that if I want to set a single layout for my site, I can propably remove most of the $isPage and $isCpct conditions.
